sdk/lib/cmlib/cmheal.c

```diff
@@ -0,0 +1,913 @@
+                      &FastIndex->List[Index + 1],
+                      MoveCount * sizeof(CM_INDEX));
+    }
+
+    /* Update the fast index count if asked */
+    if (UpdateCount)
+        FastIndex->Count--;
+}
+
+/**
+ * @brief
+ * Removes a cell from a normal key index.
+ *
+ * @param[in,out] KeyIndex
+ * The key index where a cell has to be removed.
+ *
+ * @param[in] Index
+ * The index which points to the location of the
+ * cell that is to be removed.
+ *
+ * @return
+ * Nothing.
+ */
+VOID
+CmpRemoveIndexKeyCell(
+    _Inout_ PCM_KEY_INDEX KeyIndex,
+    _In_ ULONG Index)
+{
+    ULONG MoveCount;
+    ASSERT(Index < KeyIndex->Count);
+
+    /* Calculate the number of trailing cells */
+    MoveCount = KeyIndex->Count - Index - 1;
+    if (MoveCount != 0)
+    {
+        /* Remove the cell now by moving one location ahead */
+        RtlMoveMemory(&KeyIndex->List[Index],
+                      &KeyIndex->List[Index + 1],
+                      MoveCount * sizeof(HCELL_INDEX));
+    }
+
+    /* Update the key index count */
+    KeyIndex->Count--;
+}
+
+/**
+ * @brief
+ * Removes a cell from a key value list node.
+ *
+ * @param[in,out] ValueListNode
+ * The value list node which is used by the
+ * function to update the value list count.
+ *
+ * @param[in,out] ValueListData
+ * The value list data of which a cell has to be removed.
+ *
+ * @param[in] Index
+ * The index which points to the location of the
+ * cell that is to be removed.
+ *
+ * @return
+ * Nothing.
+ */
+VOID
+CmpRemoveValueFromValueList(
+    _Inout_ PCM_KEY_NODE ValueListNode,
+    _Inout_ PCELL_DATA ValueListData,
+    _In_ ULONG Index)
+{
+    ULONG MoveCount;
+    ASSERT(Index < ValueListNode->ValueList.Count);
+
+    /* Calculate the number of trailing values */
+    MoveCount = ValueListNode->ValueList.Count - Index - 1;
+    if (MoveCount != 0)
+    {
+        /* Remove the value now by moving one location ahead */
+        RtlMoveMemory(&ValueListData->u.KeyList[Index],
+                      &ValueListData->u.KeyList[Index + 1],
+                      MoveCount * sizeof(HCELL_INDEX));
+    }
+
+    /* Update the value list count */
+    ValueListNode->ValueList.Count--;
+}
+
+/**
+ * @brief
+ * Removes the offending subkey from a root index.
+ *
+ * @param[in] Hive
+ * A pointer to a hive.
+ *
+ * @param[in] RootIndex
+ * The root index where a leaf is obtained from. Such
+ * leaf is used to check deep down the leaf for the offending
+ * subkey.
+ *
+ * @param[in] TargetKey
+ * The offending target subkey to be removed.
+ *
+ * @return
+ * Returns TRUE if the function successfully removed the target
+ * key, FALSE otherwise.
+ */
+BOOLEAN
+CmpRemoveSubkeyInRoot(
+    _In_ PHHIVE Hive,
+    _In_ PCM_KEY_INDEX RootIndex,
+    _In_ HCELL_INDEX TargetKey)
+{
+    PCM_KEY_INDEX Leaf;
+    PCM_KEY_FAST_INDEX FastIndex;
+    HCELL_INDEX LeafCell;
+    ULONG RootCountIndex;
+    ULONG LeafCountIndex;
+
+    PAGED_CODE();
+
+    /* Sanity check */
```

Comment:
Have seen that exact kind of comment several times. Superfluous. An assert at a non-empty function: What else than a sanity check should that be?